### PR TITLE
Optimize code generation of CheckRefs.

### PR DESCRIPTION
**Pull request by cmuellner**
* Remove not needed truncation
 * Remove masking, when not needed

This brought a performance boost of 6.7% in
OSS-performance's mediawiki benchmark (comparing the average
transaction rate of 4 subsequent runs each).

No regressions, when running the "all" test set on ARM64.

Example from t-test (vm-perf):

Before:
    (14) CheckRefs<0,255,192> t2:Func, 64 -> B10<Unlikely>
        Main:
              0x24c0c9d0  f9402400              ldr x0, [x0, #72]
              //movtqb w0 <- w0
              0x24c0c9d4  53001c00              uxtb w0, w0
              //andbi w0 <- w0, 255
              0x24c0c9d8  12800001              movn w1, #0x0
              0x24c0c9dc  0a010000              and w0, w0, w1
              //cmpi w0, 192
              0x24c0c9e0  53001c00              uxtb w0, w0
              0x24c0c9e4  128007e1              movn w1, #0x3f
              0x24c0c9e8  6b01001f              cmp w0, w1

              0x24c0c9ec  540000a0              b.eq #+0x14
              0x24c0c9f0  58000052              ldr x18, pc+8 (addr 0x24c0c9f8)
              0x24c0c9f4  d61f0240              br x18
              0x24c0c9f8  28c03948              ldp w8, w14, [x10]
              0x24c0c9fc  00000000              unallocated
(Unallocated)
After:
    (14) CheckRefs<0,255,192> t2:Func, 64 -> B10<Unlikely>
        Main:
              0x1660c9b4  f9402400              ldr x0, [x0, #72]
              //cmpi w0, 192
              0x1660c9b8  53001c00              uxtb w0, w0
              0x1660c9bc  128007e1              movn w1, #0x3f
              0x1660c9c0  6b01001f              cmp w0, w1
              0x1660c9c4  540000a0              b.eq #+0x14
              0x1660c9c8  58000052              ldr x18, pc+8 (addr 0x1660c9d0)
              0x1660c9cc  d61f0240              br x18
              0x1660c9d0  1a603948              unallocated (Unallocated)
              0x1660c9d4  00000000              unallocated (Unallocated)

Signed-off-by: Christoph Muellner <christoph.muellner@theobroma-systems.com>

### CODE REVIEW

**Comment by cmuellner** [comment]
@mxw : please review
@dave-estes @jim-saxman @swalk-cavium : please test if possible
Notification also for @ptomsich

**Comment by swalk-cavium** [comment]
Branch builds for me (got past the ocaml stuff first try).  Starting MOP with 6 option sets.

File: hphp/runtime/vm/jit/irlower-call.cpp [hunk 1]
```diff
@@ -477,16 +477,24 @@ void cgCheckRefs(IRLS& env, const IRInstruction* inst)  {
...
 
-      if (mask64 <= 0xff && vals64 <= 0xff) {
-        v << movtqb{bits, truncBits};
-        v << andbi{(int8_t)mask64, truncBits, maskedBits, v.makeReg()};
+      if (mask64 < 0xff && vals64 <= 0xff) {
+        v << andbi{(int8_t)mask64, bits, maskedBits, v.makeReg()};
         v << cmpbi{(int8_t)vals64, maskedBits, sf};
+      } else if (mask64 == 0xff && vals64 <= 0xff) {
+        v << cmpbi{(int8_t)vals64, bits, sf};
       } else if (mask64 <= 0xffffffff && vals64 <= 0xffffffff) {
```
Comment (dave-estes) [inline hphp/runtime/vm/jit/irlower-call.cpp]: Is this supposed to be changed to `} else if (mask64 < 0xffffffff && vals64 <= 0xffffffff) {` now, similar to the first change?

Comment (cmuellner) [inline hphp/runtime/vm/jit/irlower-call.cpp]: Correct. Thanks!

**Comment by cmuellner** [comment]
@racardoso would it be possible to test this PR on PPC64?
@mxw I am currently testing on x86_64 and will report any regressions.

**Comment by cmuellner** [comment]
@mxw I was already thinking about such an optimization. Your response encourages me to realize this.
But I think that PR should be applied independent of that. Manually generating slow code for no reason, just to have it optimized in a later pass, makes the code at least less readable and JIT times longer.

**Comment by racardoso** [comment]
@cmuellner I tested it here on PPC64 don't notice any regressions. I still testing oss-performance to see if there's any performance issues but I guess it's fine.

**Comment by mxw** [comment]
@cmuellner—I disagree.  Local vasm microoptimizations should be done in independent optimization passes (e.g., vasm-simplify).  The hhir-to-vasm lowering step shouldn't have to concern itself with lower-level optimizations.

Feel free to repurpose this PR for a vasm-simplify optimization, or create a new one; but I don't want to pollute `irlower` with one-off special cases of optimizations that belong elsewhere.

**Comment by mxw** [comment]
And I definitely disagree that it makes the code more readable this way... the logic here is "mask and compare".  The fact that when your mask is trivial, you can omit the mask step, isn't interesting when you're just trying to answer the question, "What does this HHIR opcode do?"

**Comment by cmuellner** [comment]
It turned out, that this patch does not work on x86_64.

Here are the additionally failing tests on x86_64 with the patch:

```
cat hphp/test/slow/u_converter/2135.php.diff
cat hphp/test/slow/u_converter/2136.php.diff
cat hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback.php.diff
cat hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback_return.php.diff
```

Output of diffs:
```
cm@cm:~/src/hhvm (hhvm-arm64-next)$ cat hphp/test/slow/u_converter/2135.php.diff
022- fromUCallback(REASON_CLOSE, ...)

cm@cm:~/src/hhvm (hhvm-arm64-next)$ cat hphp/test/slow/u_converter/2136.php.diff
Test failed because the process wrote on stderr:
sh: line 1: 30913 Segmentation fault      /usr/bin/timeout 300 ./hphp/hhvm/hhvm -c /home/cm/src/hhvm/hphp/test/slow/config.ini -vEval.EnableArgsInBacktraces=true -vEval.EnableIntrinsicsExtension=true -vRepo.Local.Mode=-- -vRepo.Central.Path=/home/cm/src/hhvm/hphp/hhvm/verify.hhbc -vEval.Jit=true -vEval.ProfileHWEnable=false -vResourceLimit.CoreFileSize=0 --file 'hphp/test/slow/u_converter/2136.php' 2>&1

cm@cm:~/src/hhvm (hhvm-arm64-next)$ cat hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback.php.diff
Test failed because the process wrote on stderr:
sh: line 1: 19536 Segmentation fault      /usr/bin/timeout 300 ./hphp/hhvm/hhvm -c /home/cm/src/hhvm/hphp/test/zend/config.ini -vEval.EnableArgsInBacktraces=true -vEval.EnableIntrinsicsExtension=true -vRepo.Local.Mode=-- -vRepo.Central.Path=/home/cm/src/hhvm/hphp/hhvm/verify.hhbc -vEval.Jit=true -vEval.ProfileHWEnable=false -vResourceLimit.CoreFileSize=0 --file 'hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback.php' 2>&1

cm@cm:~/src/hhvm (hhvm-arm64-next)$ cat hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback_return.php.diff
Test failed because the process wrote on stderr:
sh: line 1: 16165 Segmentation fault      /usr/bin/timeout 300 ./hphp/hhvm/hhvm -c /home/cm/src/hhvm/hphp/test/zend/config.ini -vEval.EnableArgsInBacktraces=true -vEval.EnableIntrinsicsExtension=true -vRepo.Local.Mode=-- -vRepo.Central.Path=/home/cm/src/hhvm/hphp/hhvm/verify.hhbc -vEval.Jit=true -vEval.ProfileHWEnable=false -vResourceLimit.CoreFileSize=0 --file 'hphp/test/zend/good/ext/intl/tests/uconverter_oop_callback_return.php' 2>&1
```

I guess the reason is, that the patch removed the truncation, which is only
ok for ARM64 since it is done there also as part of the cmpbi lowering.

My plan is now to keep this PR open until I come up with the following patches
(thanks @mxw for the hints):

 1) simplify and[b|l]i [0xff|0xffffffff] ; cmp[b|l]i -> cmp[b|l]i (already in testing)
 2) Do proper loading in cgCheckRefs to eliminate truncation (enqueued for testing)

These two should achieve at least the same result as this PR.
Additionally I'll try to come up with a generalization of 2:

 3) loadx ; movtxz ; doity -> loady ; doity

And as a last point I want to:

 4) vasm-arm: eliminate truncation for cmpbi and friends

**Comment by mxw** [comment]
@cmuellner—That plan sounds great; we can leave this PR here until those are up.

**Comment by mxw** [comment]
Closing this; the replacement PRs all seem to be up.

**Pull request closed without merging**